nox/manifest.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2017 Jon Wayne Parrott
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+import copy
+import itertools
+
+from nox._parametrize import generate_calls
+from nox.logger import logger
+from nox.sessions import Session
+
+
+class Manifest(object):
+    """Session manifest.
+
+    The session manifest provides the final list of sessions that should be
+    run by nox.
+
+    It is possible for this to be mutated during execution. This allows for
+    useful use cases, such as for one session to "notify" another.
```

Comment:
So, it's not _quite_ a chain. For instance, say `unit_tests(interpreter='2.7')` and `unit_tests(interpreter='3.6')` _both_ add the `cover` session to the manifest. The ultimate idea for me is that cover (1) gets added to the very end (e.g. it does not run immediately after the first thing to add it), and (2) gets added once.

The terminology I am using here is [borrowed from Ansible](http://docs.ansible.com/ansible/latest/playbooks_intro.html#handlers-running-operations-on-change), and the best term I could think of. I think `chain` is wrong -- it implies a slightly different thing to me (run this session _next_). But maybe there is a different term that would be better?

It is worth noting that there is no reason why "notify" and "chain" (as I am defining the terms in this comment) could not _both_ be implemented. There are valid use cases for each.